### PR TITLE
Add single data table component for all pages

### PR DESCRIPTION
**Pull request by thedev966**
## Changes
- Remove individual data table components for all pages
- Add one single data table component for all pages

## Reference
https://trello.com/c/uWJwm3Dh/77-task-features-datatable

### CODE REVIEW

File: src/components/DataTable.vue [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+      <template #default="table">
+        <div class="action-icons">
+          <i
+            class="fa-solid fa-pencil icon"
+            @click="
+              () => {
+                editHandler(table.row);
+                openAppModal();
+              }
+            "
```
Comment (josip2312) [inline src/components/DataTable.vue:22]: This should work
```suggestion
            @click="
                editHandler(table.row);
                openAppModal();
            "
```

File: src/components/DataTable.vue [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+    },
+    editHandler: {
+      type: Function,
+      required: true,
+    },
+    deleteHandler: {
+      type: Function,
+      required: true,
+    },
+  },
```
Comment (josip2312) [inline src/components/DataTable.vue:65]: This is not the vue way to do things, you should emit events on clicking these icons, and react to these event appropriately in parent components

File: src/components/DataTable.vue [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+  methods: {
+    ...mapMutations({ openAppModal: OPEN_APP_MODAL }),
+    handleCellClick: (row, column, cell) => {
```
Comment (josip2312) [inline src/components/DataTable.vue:68]: Clicking on edit or delete should not trigger the row expansion

**Review by josip2312 — CHANGES_REQUESTED** [review]

File: src/components/DataTable.vue [hunk 4]
```diff
@@ -15,13 +15,21 @@
           <i
```
Comment (josip2312) [inline src/components/DataTable.vue:15]: I just realized this is just an icon, it should be wrapped with a button

**Review by scokljat — APPROVED** [review]

File: src/components/DataTable.vue [hunk 5]
```diff
@@ -15,13 +15,21 @@
...
             class="fa-solid fa-pencil icon"
             @click="
-              $emit('edit', table.row);
-              openAppModal();
+              (event) => {
```
Comment (josip2312) [inline src/components/DataTable.vue:18]: You can use one function here and pass it `(table.row, $event)`, so the template is cleaner

File: src/components/DataTable.vue [hunk 6]
```diff
@@ -15,13 +15,21 @@
...
+                $emit('edit', table.row);
+                openAppModal();
+                preventRowExpansion(event);
+              }
             "
           ></i>
           <i
             class="fa-solid fa-trash icon"
-            @click="$emit('delete', table.row)"
+            @click="
```
Comment (josip2312) [inline src/components/DataTable.vue:27]: same as above

File: src/components/DataTable.vue [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+  methods: {
+    ...mapMutations({ openAppModal: OPEN_APP_MODAL }),
+    handleCellClick: (row, column, cell) => {
+      cell.parentElement.classList.toggle('expanded');
+    },
+    preventRowExpansion: (event) => {
+      event.preventDefault();
+      event.stopPropagation();
+      event.stopImmediatePropagation();
+    },
```
Comment (josip2312) [inline src/components/DataTable.vue:78]: Is there a reason you used all three of these?

File: src/components/DataTable.vue [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+    tableColumns: {
+      type: Array,
+      required: true,
+    },
+  },
+  emits: ['edit', 'delete', 'pageChange'],
+  methods: {
+    ...mapMutations({ openAppModal: OPEN_APP_MODAL }),
+    handleCellClick: (row, column, cell) => {
```
Comment (josip2312) [inline src/components/DataTable.vue:71]: Vue method should never be an arrow function, try logging `this` inside of it
```suggestion
    handleCellClick: (row, column, cell) => {
```

**Review by josip2312 — CHANGES_REQUESTED** [review]

File: src/components/DataTable.vue (hunk 8 above)
Comment (thedev966) [reply src/components/DataTable.vue]: I figured it out. Fixed.

File: src/components/DataTable.vue [hunk 9]
```diff
@@ -68,13 +67,16 @@ export default {
...
+    handleCellClick(row, column, cell) {
       cell.parentElement.classList.toggle('expanded');
     },
-    preventRowExpansion: (event) => {
-      event.preventDefault();
+    preventRowExpansion(event) {
       event.stopPropagation();
-      event.stopImmediatePropagation();
+    },
+    handleActionButtons(e, emitEvent, row) {
```
Comment (josip2312) [inline src/components/DataTable.vue:76]: Rather seperate this to `handleEditAction` and `handleDeleteAction`

File: src/components/DataTable.vue [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+      required: true,
+    },
+  },
+  emits: ['edit', 'delete', 'pageChange'],
+  methods: {
+    ...mapMutations({ openAppModal: OPEN_APP_MODAL }),
+    handleCellClick(row, column, cell) {
+      cell.parentElement.classList.toggle('expanded');
+    },
+    preventRowExpansion(event) {
```
[the remark this answers is not in the document]
Comment (thedev966) [reply src/components/DataTable.vue:73]: FIxed.

File: src/components/DataTable.vue [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+    handleCellClick(row, column, cell) {
+      cell.parentElement.classList.toggle('expanded');
+    },
+    handleEditAction(row) {
+      this.$emit('edit', row);
+      this.openAppModal();
+    },
+    handleDeleteAction(row) {
+      this.$emit('delete', row);
+      this.openAppModal();
```
Comment (josip2312) [inline src/components/DataTable.vue:67]:
```suggestion
```

Comment (josip2312) [reply src/components/DataTable.vue:67]: FIx this and you can merge

**Pull request merged**